Finish migration from SqlEngine to Jdbi

jonbarksdale opened the pull request:
--
Remove SqlEngine, replace with Jdbi

Last bit of SqlEngine replaced with Jdbi.  The structure of the
Tenant_data table is not very friendly to Jdbi, so it's a bit messier
than the other migrations though.

As part of this, the SqlEngine instance was also removed from the
BaseSQLTest, though it appeared to be unused anyway.

This completes issue #4

-- 


build passes, checkstyle seems satisfied.

wolpert commented:
Looks like this will also complete https://github.com/wolpert/svarm/issues/20

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -52,28 +56,27 @@ public class V1SingleEntryEngine implements TableDefinitionEngine {
                              final JsonEngine jsonEngine) {
+    this.dataSourceManager = dataSourceManager;
     this.jsonEngine = jsonEngine;
-    LOGGER.info("V1SingleEntryEngine({},{})", metrics, sqlEngine);
```
Just replace the sqlEngine with the data source manager. Probably should also include the jsonEngine. (Cannot suggest for deleted lines)

`LOGGER.info("V1SingleEntryEngine({},{},{})", metrics, dataSourceManager, jsonEngine);`

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -52,28 +56,27 @@ public class V1SingleEntryEngine implements TableDefinitionEngine {
 
   /**
-   * Uses the murmur3/32b hashing function.
+   * Returns a hash function supplier, currently the murmur3/32b hashing function.
```
Need a new story.. I moved the hashing function to a common class and standardized on murmur3. No need for table-specific implementations. (No change here, just making a new issue for this)

wolpert replied:
https://github.com/wolpert/svarm/issues/43

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -198,18 +166,55 @@ public void write(final TenantTable tenantTable, final EntryInfo entryInfo) {
-          }
-        });
+
+    int updateCount = Jdbi.create(dataSourceManager.getDataSource(tenantTable))
```
```suggestion
    final int updateCount = Jdbi.create(dataSourceManager.getDataSource(tenantTable))
```

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -97,52 +100,17 @@ public String definitionName() {
-          }
-        });
+
+    return Jdbi.create(dataSourceManager.getDataSource(tenantTable))
```
I'm not clear if by using the static method Jdbi.create() you get all the build stuff you see in the DataSourceModule... the immutibles stuff should fail.

At least we should pull in that jdbi instance to this class, and use that. (Here and below)

wolpert replied:
To add more context, we will want things like metrics... example call I want to enable. 
`
   jdbi.setSqlLogger(new InstrumentedSqlLogger(metricRegistry));
`
I'm guessing we'll need to cache them with their data sources? (Instead of caching the datasource, we cache the jdbi? As long as we can delete it that is)

jonbarksdale replied:
I can see that, but I think it makes more sense to add a `getJdbi` method to the DataSourceManager, allow us to cache it and instrument construction as necessary.  I think the Jdbi interface is actually quite cheap to construct since the datasource should manage caching/pooling connections. 

Probably a task for another issue though.

wolpert replied:
We can add it in later, for sure. But note we are actually going to configure the JDBI instances... so not sure its as cheap as you think.

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -198,18 +166,55 @@ public void write(final TenantTable tenantTable, final EntryInfo entryInfo) {
+    return result;
+  }
+
+  private class EntryInfoResultSetScanner implements ResultSetScanner<Optional<EntryInfo>> {
```
I have the feeling this can be simplified. It's with parity of what I had before. (Assuming the tests work) No need to change here. But this probably can be redone in a future  PR.

wolpert replied:
I updated https://github.com/wolpert/svarm/issues/11 to include this simplification.

wolpert reviewed (changes requested):
I assume all the tests run fine? 
`gradle clean build test integ -Pe2e` (Though note sometimes the e2e fails the first time due to overloading the computer. Run again without the clean if that happens)

Made a suggestion on the Jdbi.create() -> jdbi.create().

wolpert commented:
FYI, I pulled the PR and see an error when running the test... we should see if we can reproduce this with a test in the dao for it.

```
14:07:20.797 [dw-34 - GET /v1/tenant/TableTenantLifecycleTest.tenant/table/TableTenantLifecycleTest.table/entry/newEntry] [trace:TableTenantLifecycleTest.createTable] ERROR i.d.j.errors.LoggingExceptionMapper - Error handling a request: 3c64eccf809de120
org.jdbi.v3.core.result.ResultSetException: Error reading result set [statement:"select * from TENANT_DATA where ID = :id", arguments:{positional:{}, named:{id:newEntry}, finder:[]}]
	at org.jdbi.v3.core.result.ResultBearing$1.scanResultSet(ResultBearing.java:60)
	at org.jdbi.v3.core.statement.Query.scanResultSet(Query.java:61)
	at org.svarm.node.engine.impl.V1SingleEntryEngine.lambda$read$0(V1SingleEntryEngine.java:113)
	at org.jdbi.v3.core.Jdbi.withHandle(Jdbi.java:367)
	at org.svarm.node.engine.impl.V1SingleEntryEngine.read(V1SingleEntryEngine.java:110)
	at org.svarm.node.manager.TenantTableEntryManager.read(TenantTableEntryManager.java:68)
	at org.svarm.node.resource.TenantTableEntryResource.readTenantTableEntry(TenantTableEntryResource.java:71)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:134)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:177)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:81)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:235)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1665)
	at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:36)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
	at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:46)
	at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:40)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:527)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1380)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at io.dropwizard.metrics.jetty11.InstrumentedHandler.handle(InstrumentedHandler.java:307)
	at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:822)
	at io.dropwizard.jetty.ZipExceptionHandlingGzipHandler.handle(ZipExceptionHandlingGzipHandler.java:26)
	at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:46)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:416)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:385)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:272)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:934)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1078)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.sql.SQLException: invalid cursor state: identifier cursor not positioned on row in UPDATE, DELETE, SET, or GET statement: ; ResultSet is positioned after last row
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCResultSet.getCurrent(Unknown Source)
	at org.hsqldb.jdbc.JDBCResultSet.getColumnValue(Unknown Source)
	at org.hsqldb.jdbc.JDBCResultSet.getColumnInType(Unknown Source)
	at org.hsqldb.jdbc.JDBCResultSet.getInt(Unknown Source)
	at org.hsqldb.jdbc.JDBCResultSet.getInt(Unknown Source)
	at com.mchange.v2.c3p0.impl.NewProxyResultSet.getInt(NewProxyResultSet.java:451)
	at org.svarm.node.engine.impl.V1SingleEntryEngine$EntryInfoResultSetScanner.scanResultSet(V1SingleEntryEngine.java:207)
	at org.svarm.node.engine.impl.V1SingleEntryEngine$EntryInfoResultSetScanner.scanResultSet(V1SingleEntryEngine.java:182)
	at org.jdbi.v3.core.result.ResultBearing$1.scanResultSet(ResultBearing.java:58)
	... 73 common frames omitted
Caused by: org.hsqldb.HsqlException: invalid cursor state: identifier cursor not positioned on row in UPDATE, DELETE, SET, or GET statement: ; ResultSet is positioned after last row
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	... 83 common frames omitted
14:07:20.802 [dw-34 - GET /v1/tenant/TableTenantLifecycleTest.tenant/table/TableTenantLifecycleTest.table/entry/newEntry] [trace:TableTenantLifecycleTest.createTable] TRACE o.s.s.resource.TraceUuidResource - filter(response):TableTenantLifecycleTest.createTable
14:07:20.881 [vert.x-eventloop-thread-0] [trace:] TRACE o.s.common.config.engine.WatchEngine - node/65c81317-6db2-4889-8505-2247cd78ddcd:watchResponse(header {
```

wolpert commented:
Welp, the V1SingleEntryEngine doesn't have a test case. We should create one and have it extend BaseSQLTest to get the components enabled.

wolpert commented on java/node/src/main/java/org/svarm/node/engine/impl/V1SingleEntryEngine.java:
```diff
@@ -154,37 +122,37 @@ public Optional<EntryInfo> read(final TenantTable tenantTable, final String enti
+
+            String dataType;
+
+            if (element.isNumber()) {
+              dataType = INTEGER_TYPE;
+            } else {
+              if (element.isTextual()) {
+                dataType = STRING_TYPE;
+              } else {
+                throw new IllegalArgumentException("Unknown type: " + element.getNodeType());
               }
```
I couldn't do the suggestion thing due to the deleted line... but how about
```
if () {
...
} else if () {
...
} else {
  throw ...
}

jonbarksdale commented:
Updated with suggestions from review, rebased on top of other changes, updated the test for the changes and fixed the bug in the code that was failing the test.
